server/src/chunkworker.rs

```diff
@@ -72,12 +75,18 @@ struct ChunkWorker {
 
     /// A map of currently open region files
     open_regions: HashMap<RegionPosition, RegionFile>,
+
+    /// World generator for new chunks.
+    world_generator: Arc<dyn WorldGenerator>,
 }
 
 /// Starts a chunk worker on a new thread.
 /// The returned channels can be used
 /// to communicate with the worker.
-pub fn start(world_dir: &str) -> (Sender<Request>, Receiver<Reply>) {
+pub fn start(
+    world_dir: &str,
+    world_gen: Arc<dyn WorldGenerator>,
```

Comment:
So I'm trying to figure out the correct way to pass a `WorldGenerator` instance all the way from the `main.rs` to the `chunkworker.rs`. I see that `chunklogic.rs` is setting-up chunkworker, but I don't know how I would access a component from there. Any pointers?

Reply:
Yes, this is an annoying caveat of Specs. To initialize the `ChunkWorkerHandle` resource, Specs calls `ChunkWorkerHandle::default` when `Dispatcher::setup` is called. There's no way to access components from the `default` function.

There is a solution in the Specs book [here](https://slide-rs.github.io/specs/04_resources.html#default-for-resources). You should likely use the second solution, which is to use `ReadExpect<'a, ChunkWorkerHandle>` rather than `Read` and then insert the chunk worker handle into the world in `ChunkLoadSystem::setup`. Here, you can access any components and resources you need for the chunk worker.

Reply:
So if I understand correctly, using a `ReadExpect` removes the need for the Default trait on the `ChunkWorkerHandle`. However, the `chunkworker` is also initialized a bit further down: https://github.com/caelunshun/feather/blob/01d6d08fc1bcadc3788e375e1bb380ce6ffaed03/server/src/chunk_logic.rs#L83-L91

I'm still a bit unclear on how I can access a component from there. I can add a `Read<Arc<dyn WorldGenerator>>` to the `SystemData` type, but I don't know how to unpack it from the `setup` function (with only `&mut world` provided).

Perhaps this sort of component dependency is not a "correct" way of tackling this?

Reply:
Yes, your understanding of `ReadExpect` is correct.

You can access resources from the world using `World::fetch` and `World::fetch_mut`. To get the level data, for example:
```rust
let level = world.fetch_mut::<LevelData>();
```